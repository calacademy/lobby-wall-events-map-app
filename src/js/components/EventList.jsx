'use strict'

var React = require('react')
var $ = require('jquery')

// Displays an event for each event list member
var Event = React.createClass({

  displayName: "Event",

  _convertMilitaryTime: function(str) {
    var strTimes = ''
    var hours24 = parseInt(str.substring(0, 2),10)
    var hours = ((hours24 + 11) % 12) + 1
    var amPm = hours24 > 11 ? 'PM' : 'AM'
    var minutes = str.substring(2)
    strTimes += hours + ':' + minutes + ' ' + amPm
    return strTimes
  },

  render: function render() {

    if ((this.props.dataset.title === "Museum Opens") || (this.props.dataset.title === "Museum Closes")) {

      return (
        <article className={(this.props.time < this.props.currTime) ? 'event past' : 'event'}>
          <h2 className="notranslate">{this._convertMilitaryTime(this.props.time)}</h2>
          <h1 className="open-close">{this.props.dataset.title}</h1>
        </article>
      )

    } else {

      if (this.props.time < this.props.closingTime.replace(/:/, '')) {

        var markup = {__html: this.props.dataset.body}

        return (
          <article className={(this.props.time < this.props.currTime) ? 'event past' : 'event'}>
            <h2 className="notranslate">{this._convertMilitaryTime(this.props.time)}</h2>
            <h1>{this.props.dataset.title}</h1>
            <span className="location">{this.props.dataset.location}</span>
            <div dangerouslySetInnerHTML={markup} />
          </article>
        )

      } else {

        return null

      }

    }

  }

})

// Displays an event-list for each timeblock
var EventListTimeBlock = React.createClass({

  displayName: "EventListTimeBlock",

  render: function render() {

    var rows = []

    this.props.dataset.events.forEach(function(dataset) {
      rows.push(<Event dataset={dataset} key={dataset.nid} time={this.props.dataset.time} currTime={this.props.currTime} closingTime={this.props.closingTime} />)
    }.bind(this))

    return (
      <section className="time-block">
        {rows}
      </section>
    )
  }

})

// Displays events
var EventList = React.createClass({

  displayName: "EventList",

  getInitialState: function() {
    return {
      info: null,
      showBackup: false,
      currentTime: null,
      nextPage: false,
      prevPage: false,
      eventPage: 'first',
    };
  },

  _setCurrentTime: function() {
    var date = new Date()
    var hours = date.getHours()
    var minutes = date.getMinutes()
    if (hours < 10) hours = "0" + hours.toString()
    if (minutes < 10) minutes = "0" + minutes.toString()
    var time = hours + '' + minutes
    this.setState ({
      currentTime: time
    })
  },

  _getNiceTodayDate: function() {

    var monthNames = [
      "January", "February", "March",
      "April", "May", "June", "July",
      "August", "September", "October",
      "November", "December"
    ]
    var date = new Date()
    var day = date.getDate()
    var monthIndex = date.getMonth()
    var year = date.getFullYear()

    var sfx = ["th","st","nd","rd"]
    var val = day%100;
    var daySfx = (sfx[(val-20)%10] || sfx[val] || sfx[0])

    return (monthNames[monthIndex] + ' ' + day + daySfx + ', ' + year)

  },

  _parseEventsByTime: function(data) {
    if (data) {
      // get all the timeslots in current data
      var times = []
      data.forEach(function(dataset) {
        if(dataset.time_slots) {
          if (Array.isArray(dataset.time_slots)) {
            dataset.time_slots.forEach(function(ts) {
              times.push(ts.replace(/:/, ''))
            })
          }
        }
      })
      times.sort()
      var uniqueTimes = []
      $.each(times, function(i, el){
        if($.inArray(el, uniqueTimes) === -1) uniqueTimes.push(el)
      })
      times = uniqueTimes
      // from times create array of time blocks - each member having
      // object comprised of time:string and event-list:array
      var timeBlocks = []
      times.forEach(function(time) {
        var obj = {
          'time': time,
          'events': []
        }
        timeBlocks.push(obj)
      })
      // iterate thru current data and drop event info into matching
      // time block member object event-lists based on data timeslots
      data.forEach(function(dataset) {
        if(dataset.time_slots) {
          if (Array.isArray(dataset.time_slots)) {
            dataset.time_slots.forEach(function(ts) {
              var time = ts.replace(/:/, '')
              var obj = {
                'nid': dataset.nid,
                'location': dataset.location,
                'title': dataset.title,
                'body': dataset.body
              }
              timeBlocks.forEach(function(tb) {
                if (tb.time === time) {
                  tb.events.push(obj)
                }
              })
            })
          }
        }
      })
      this.setState ({
        info: timeBlocks
      })
    }
  },

  _multiPageCheck: function() {
    var more = false
    var limit = $('#event-list-container').width() * 2
    $('article').each(function () {
      var offset = $(this).offset();
      if (offset.left > limit) {
        more = true
      }
		})
    if (this.state.eventPage === 'last') {
      more = false
    }
    this.setState({
      nextPage: more
    })
  },

  _handleClickNext: function() {
    var adj = 0 - $('#event-list-container').width()
    $('.row-wrapper').css('margin-left',adj)
    this.setState({
      nextPage: false,
      prevPage: true,
      eventPage: 'last'
    })
  },

  _handleClickPrev: function() {
    $('.row-wrapper').css('margin-left','0')
    this.setState({
      nextPage: true,
      prevPage: false,
      eventPage: 'first'
    })
  },

  componentWillReceiveProps: function(nextProps) {

    this._parseEventsByTime(nextProps.dataset)

    if (nextProps.networkConnected === false) {
      this.setState({
        showBackup: true
      })
    } else {
      this.setState({
        showBackup: false
      })
    }

    this._multiPageCheck()

  },

  componentDidMount: function() {
    // set currentTime every 5 minutes
    this._setCurrentTime()
    setInterval(()=>this._setCurrentTime(), (60000 * 2))
  },

  render: function render() {

    var classViewExpand = ''
    if (this.props.datacount <= 13) {
      classViewExpand = 'halfsize'
    } else {
      classViewExpand = 'fullsize'
    }

    if (this.state.info) {

      var rows = []

      this.state.info.forEach(function(dataset) {
        rows.push(<EventListTimeBlock dataset={dataset} key={dataset.time} currTime={this.state.currentTime} closingTime={this.props.closingTime} />)
      }.bind(this));

      return (
        <div>
          <div id="event-list-container" className={this.state.showBackup ? 'hide' : ''}>
            <div className="event-list-header">
              <span className="whats-on-today">{'What\u2019s on Today'}</span>
              <span>{this._getNiceTodayDate()}</span>
            </div>
            <div className={classViewExpand + ' row-wrapper'}>
              <div className="outer-container">{rows}</div>
            </div>
          </div>
          <div id="event-list-container-backup" className={this.state.showBackup ? 'notranslate' : 'hide notranslate'}>
            <div className="event-list-header">
              <span className="whats-on-today">{'What\u2019s on Today'}</span>
              <span>{this._getNiceTodayDate()}</span>
            </div>
            <div className={classViewExpand + ' row-wrapper'}>
              <div className="outer-container">{rows}</div>
            </div>
          </div>
          <div className={((this.state.nextPage) && (this.props.appOnPage === 'events')) ? 'pager next' : 'hide pager next'} onClick={this._handleClickNext}></div>
          <div className={((this.state.prevPage) && (this.props.appOnPage === 'events')) ? 'pager prev' : 'hide pager prev'} onClick={this._handleClickPrev}></div>
        </div>
      )

    } else {

      return (
        <div className={this.props.hide ? 'hide' : ''}>
          <div id="event-list-container">
            <div className="event-list-header">
              <span className="whats-on-today">Information unavailable</span>
            </div>
          </div>
        </div>
      )

    }

  }

})

export default EventList
