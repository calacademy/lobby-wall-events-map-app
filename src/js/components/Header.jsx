'use strict'

var React = require('react')
var ReactDOM = require('react-dom')

import Translator from './Translator.jsx'
import Greeting from './Greeting.jsx'

var Header = React.createClass({

  displayName: "Header",

  render: function() {

    return (
      <div className={this.props.hide ? 'hide' : ''}>
        <header className="header" role="contentinfo">
          <div className={!this.props.networkConnected ? 'hide' : ''}>
            <Greeting />
          </div>
          <div className={!this.props.networkConnected ? 'hide' : ''}>
            <Translator networkConnected={this.props.networkConnected} appOnPage={this.props.appOnPage} />
          </div>
        </header>
      </div>
    )

  }
})

export default Header
