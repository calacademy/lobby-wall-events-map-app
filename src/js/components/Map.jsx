'use strict'

var React = require('react')
var $ = require('jquery')

// Displays events
var Map = React.createClass({

  displayName: "Map",

  getInitialState: function() {
    return {
      nextPage: true,
      prevPage: false,
      mapPage: 'first',
    };
  },

  _handleClickNext: function() {
    var adj = 0 - $('#map-container').width()
    $('.map').css('margin-left',adj)
    this.setState({
      nextPage: false,
      prevPage: true,
      mapPage: 'last'
    })
  },

  _handleClickPrev: function() {
    $('.map').css('margin-left','0')
    this.setState({
      nextPage: true,
      prevPage: false,
      mapPage: 'first'
    })
  },

  render: function render() {

    return (
      <div>
        <div id="map-container">
          <div className="map">
            <div className="map-page"></div>
            <div className="map-page"></div>
          </div>
          <div className={((this.state.nextPage) && (this.props.appOnPage === 'map')) ? 'pager next' : 'hide pager next'} onClick={this._handleClickNext}></div>
          <div className={((this.state.prevPage) && (this.props.appOnPage === 'map')) ? 'pager prev' : 'hide pager prev'} onClick={this._handleClickPrev}></div>
        </div>
      </div>
    )

  }

})

export default Map
