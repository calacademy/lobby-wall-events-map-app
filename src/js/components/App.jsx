'use strict'

var React = require('react')
var ReactDOM = require('react-dom')
var $ = require('jquery')

import Map from './Map.jsx'
import Header from './Header.jsx'
import EventList from './EventList.jsx'

var App = React.createClass({

  displayName: "App",

  getInitialState: function() {
    return {
      networkConnected: false,
      eventDataInit: false,
      eventData: null,
      eventCount: null,
      closingTime: '17:00', // default
      nextPage: true,
      prevPage: false,
      appOnPage: 'map',
    }
  },

  _testNetworkConnection: function() {
    if (navigator.onLine) {
      // only set state as needed to avoid extraneous rendering
      if (this.state.networkConnected === false) {
        this.setState({
          networkConnected: true
        })
      }
      // if init data not pulled yet, try
      if (this.state.eventDataInit === false) {
        this._getEventData()
      }
    } else {
      if (this.state.networkConnected === true) {
        this.setState({
          networkConnected: false
        })
      }
    }
  },

  _getEventData: function() {
    var closingTime = this.state.closingTime
    $.ajax({
      url: process.env.EVENTS_API_URL,
      dataType: process.env.EVENTS_API_DATATYPE,
      type: "GET",
      async: true,
      success: function(data) {
        var i = 0
        data.forEach(function(datum) {
          if (datum.time_slots) {
            i += datum.time_slots.length
          }
          if (datum.title === 'Museum Closes') {
            closingTime = datum.time_slots[0]
          }
        })
        if (i > 0) {
          this.setState ({
            eventDataInit: true,
            eventData: data,
            eventCount: i,
          })
          // request event data every hour after first successful pull
          setInterval(()=>this._updateEventData(), (60000 * 60))
        }
        this.setState ({
          closingTime: closingTime,
        })
      }.bind(this)
    })
  },

  _updateEventData: function() {
    if (navigator.onLine) {
      var closingTime = this.state.closingTime
      $.ajax({
        url: process.env.EVENTS_API_URL,
        dataType: process.env.EVENTS_API_DATATYPE,
        type: "GET",
        async: true,
        success: function(data) {
          var i = 0
          data.forEach(function(datum) {
            if (datum.time_slots) {
              i += datum.time_slots.length
            }
            if (datum.title === 'Museum Closes') {
              closingTime = datum.time_slots[0]
            }
          })
          if (i > 0) {
            this.setState ({
              eventDataInit: true,
              eventData: data,
              eventCount: i,
            })
          }
          this.setState ({
            closingTime: closingTime,
          })
        }.bind(this)
      })
    }
  },

  _handleClickNext: function() {
    var adj = 0 - $('#app-container').width()
    $('#app-container').css('margin-left',adj)
    this.setState({
      nextPage: false,
      prevPage: true,
      appOnPage: 'events',
    })
  },

  _handleClickPrev: function() {
    var adj = $('#app-container').width()
    $('#app-container').css('margin-left','0')
    this.setState({
      nextPage: true,
      prevPage: false,
      appOnPage: 'map',
    })
  },

  componentDidMount: function() {
    // establish network connection
    this._testNetworkConnection()
    // test network connectivity every 10 seconds
    setInterval(()=>this._testNetworkConnection(), 10000)
  },

  render: function render() {
    return (
      <div id='app-container'>
        <div id='app-pager-container'>
          <div id='app-page-map'>
            <Map appOnPage={this.state.appOnPage} />
          </div>
          <div id='app-page-events'>
            <Header hide={!this.state.eventDataInit} networkConnected={this.state.networkConnected} appOnPage={this.state.appOnPage} />
            <EventList hide={!this.state.eventDataInit} networkConnected={this.state.networkConnected} dataset={this.state.eventData} datacount={this.state.eventCount} closingTime={this.state.closingTime} appOnPage={this.state.appOnPage} />
            <div className={!this.state.eventDataInit ? 'no-data-init notranslate' : 'hide no-data-init notranslate'}>
              <p>Loading data. Thank you for your patience.</p>
            </div>
          </div>
        </div>
        <div className={this.state.nextPage ? 'app-pager app-next' : 'hide app-pager app-next'} onClick={this._handleClickNext}></div>
        <div className={this.state.prevPage ? 'app-pager app-prev' : 'hide app-pager app-prev'} onClick={this._handleClickPrev}></div>
      </div>
    )
  }
})

export default App
