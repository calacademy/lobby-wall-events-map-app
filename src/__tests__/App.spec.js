import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import App from '../js/components/App'
import Map from '../js/components/Map'
import Header from '../js/components/Header'
import EventList from '../js/components/EventList'

describe('App', () => {

  let app

  beforeEach(() => {
    app = shallow(<App />)
  })

  it('Default state - networkConnected is false', () => {
    expect(app.state('networkConnected')).toBeFalsy()
  })

  it('Default state - eventDataInit is false', () => {
    expect(app.state('eventDataInit')).toBeFalsy()
  })

  it('Default state - eventData is null', () => {
    expect(app.state('eventData')).toBeNull()
  })

  it('Default state - eventCount is null', () => {
    expect(app.state('eventCount')).toBeNull()
  })

  //it('_testNetworkConnection() called - networkConnected is true', () => {
    //app.instance()._testNetworkConnection()
    //expect(app.state('networkConnected')).toBeTruthy()
  //})

  it('renders one <Map /> component', () => {
    expect(app.find(Map).length).toEqual(1)
  })

  it('renders one <Header /> component', () => {
    expect(app.find(Header).length).toEqual(1)
  })

  it('renders one <EventList /> component', () => {
    expect(app.find(EventList).length).toEqual(1)
  })

  // regression test - jest snapshot
  it('Enzyme shallow matches snapshot', () => {
    expect(shallowToJson(app)).toMatchSnapshot()
  })

})
