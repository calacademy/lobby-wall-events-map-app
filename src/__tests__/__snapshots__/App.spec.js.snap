exports[`App Enzyme shallow matches snapshot 1`] = `
<div
  id="app-container">
  <div
    id="app-pager-container">
    <div
      id="app-page-map">
      <Map
        appOnPage="map" />
    </div>
    <div
      id="app-page-events">
      <Header
        appOnPage="map"
        hide={true}
        networkConnected={false} />
      <EventList
        appOnPage="map"
        closingTime="17:00"
        datacount={null}
        dataset={null}
        hide={true}
        networkConnected={false} />
      <div
        className="no-data-init notranslate">
        <p>
          Loading data. Thank you for your patience.
        </p>
      </div>
    </div>
  </div>
  <div
    className="app-pager app-next"
    onClick={[Function]} />
  <div
    className="hide app-pager app-prev"
    onClick={[Function]} />
</div>
`;
